### PR TITLE
refactor!: move autonat into separate package

### PR DESCRIPTION
**Pull request by wemeetagain**
## Description
- This is an attempt to start #1913, see what kind of issues arise from it
- Move autonat into separate package
- Move error codes and messages into @libp2p/interface/errors

## Notes & open questions
- @achingbrain  How does the "Browser <script> tag" section of the readme work? is it stale?

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by wemeetagain** [comment]
Implementation notes:
- Add the dependencies / devDependencies in the new package.json.
- Remove Components interface (and defaultComponents impl) used in tests
- Update all docs / examples
- Remove subpath exports, tsdoc entrypoints
- Transfer proto generation scripts

File: packages/interface/src/errors.ts [hunk 1]
```diff
@@ -65,3 +65,83 @@ export class InvalidCryptoTransmissionError extends Error {
 
   static readonly code = 'ERR_INVALID_CRYPTO_TRANSMISSION'
 }
+
+export enum messages {
```
Comment (wemeetagain) [inline packages/interface/src/errors.ts:69]: Maybe messages should stay in the libp2p package since they do not need to be standardized

**Comment by maschad** [comment]
Let's hold off on merging any breaking changes until https://github.com/libp2p/js-libp2p/pull/2074 is merged

**Review by maschad — CHANGES_REQUESTED** [review]
I think we should move the error codes and messages into @libp2p/interface/errors first in a separate PR

File: packages/interface/src/errors.ts (hunk 1 above)
Comment (achingbrain) [reply packages/interface/src/errors.ts:69]: I think each module should probably have it's own specific messages, otherwise we'd need to ship libp2p to fix an autonat typo, for example.

Some generic messages might live in `@libp2p/interface`.

File: packages/protocol-autonat/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    {
+      "path": "../logger"
+    },
+    {
+      "path": "../peer-id-factory"
+    },
+    {
+      "path": "../transport-tcp"
+    }
+  ]
```
Comment (achingbrain) [inline packages/protocol-autonat/tsconfig.json:35]: I think some of these might not be necessary?

**Review by achingbrain — CHANGES_REQUESTED** [review]
Comments inline.   I've targeted the v1 release branch - please can you add some copy to the migration guide to cover these breaking changes.

File: doc/migrations/v0.46-v1.0.md [hunk 3]
```diff
@@ -64,6 +64,22 @@ const libp2p = await createLibp2p({
 const keychain: KeyChain = libp2p.services.keychain
 ```
 
+### AutoNAT
```
Comment (maschad) [inline doc/migrations/v0.46-v1.0.md:67]: I think we should move all these changes to the [v0.46-v1.0.0.md](https://github.com/libp2p/js-libp2p/blob/release-v1.0/doc/migrations/v0.46-v1.0.0.md) file

**Review by maschad — CHANGES_REQUESTED** [review]
Looks good, left one comment.

File: doc/migrations/v0.46-v1.0.0.md [hunk 4]
```diff
@@ -1,16 +1,101 @@
...
-A migration guide for refactoring your application code from libp2p `v0.46` to `v1.0.0`.
+A migration guide for refactoring your application code from libp2p v0.46 to v1.0.0.
 
 ## Table of Contents <!-- omit in toc -->
 
-- [New features](#new-features)
-- [Breaking changes](#breaking-changes)
+- [API](#api)
+- [Module Updates](#module-updates)
+- [Metrics](#metrics)
```
Comment (maschad) [inline doc/migrations/v0.46-v1.0.0.md:10]:
```suggestion
- [API](#api)
- [Keychain](#keychain)
- [AutoNAT](#autonat)
- [Metrics](#metrics) 
```

File: doc/migrations/v0.46-v1.0.0.md [hunk 5]
```diff
@@ -1,16 +1,101 @@
...
+    keychain: (components) => new DefaultKeyChain(components, {
+      ...DefaultKeyChain.generateOptions()
+    })
+  }
+})
+
+const keychain: KeyChain = libp2p.services.keychain
+```
+
+### AutoNAT
```
Comment (maschad) [inline doc/migrations/v0.46-v1.0.0.md:67]:
```suggestion
## AutoNAT
```

File: doc/migrations/v0.46-v1.0.0.md [hunk 6]
```diff
@@ -1,16 +1,101 @@
+<!--Specify versions for migration below-->
 # Migrating to libp2p@1.0.0 <!-- omit in toc -->
 
-A migration guide for refactoring your application code from libp2p `v0.46` to `v1.0.0`.
+A migration guide for refactoring your application code from libp2p v0.46 to v1.0.0.
```
Comment (maschad) [inline doc/migrations/v0.46-v1.0.0.md:4]: We can revert this change.

File: doc/migrations/v0.46-v1.0.0.md (hunk 5 above)
Comment (wemeetagain) [reply doc/migrations/v0.46-v1.0.0.md:67]: These are intentionally `###` to be under `## API`

File: doc/migrations/v0.46-v1.0.0.md (hunk 6 above)
Comment (wemeetagain) [inline doc/migrations/v0.46-v1.0.0.md:4]:
```suggestion
A migration guide for refactoring your application code from libp2p `v0.46` to `v1.0.0`.
```

**Comment by wemeetagain** [comment]
Can you please take over this PR and fix up the migration guide the way you want?

**Review by maschad — APPROVED** [review]
